expo sqlite / expo-file-system unsupported url 1002 on ios bundle

### Minimal reproducible example

https://github.com/Sznapsollo/expo_ios_bundle_unsupportedurl_1002

### Summary

Case: copy sqlite db3 database from app assets to documentDirectory.

While it works perfectly on android expo, android simulator bundle, android prod bundle, ios expo
It does not work on ios bundle (simulator or production) returning error below. It happens when i build locally and also when app is build using expo services.

Exception is Error Domain=NSURLErrorDomain Code=-1002 "unsupported URL" (also attached screen from catch lot on ios)

It happens when performing the following code:
```
       await FileSystem.downloadAsync(
          Asset.fromModule(require('./assets/test_database.db3')).uri,
          FileSystem.documentDirectory + 'SQLite/test_database.db3'
        );
```
db3 extensions are added and also configured in metro.config.js in assetExts

I have browsed many docs about this but nothing works so far.
I have found similar bug report - https://github.com/expo/expo/issues/23455
This report contains info how to handle this by @alanjhughes but comment is what turns out I am actually using and the result is still 1002

![image](https://github.com/expo/expo/assets/20971560/50bf795e-082f-4766-bbae-4a8eb22b0bee)

Thank you kindly for your help

### Environment

expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.2.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.11.0 - /usr/local/bin/node
      Yarn: 1.22.21 - /usr/local/bin/yarn
      npm: 10.2.4 - /usr/local/bin/npm
      Watchman: 2023.12.04.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.14.3 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.2, iOS 17.2, macOS 14.2, tvOS 17.2, visionOS 1.0, watchOS 10.2
    IDEs:
      Xcode: 15.2/15C500b - /usr/bin/xcodebuild
    npmGlobalPackages:
      eas-cli: 6.0.0
    Expo Workflow: managed

Hi @gabrieldonadel  thank you for your response. I have already tried copyAsync but disregarded it since it was not working presenting the following error  (when running on dev expo go app) - at least downloadAync works on everything except  for production ios. Is there any working example with copyAsync?

 WARN  [Error: Call to function 'ExponentFileSystem.copyAsync' has been rejected.
→ Caused by: java.io.IOException: Location 'http://192.168.0.95:8081/assets/?unstable_path=.%2Fassets%2Ftest_database.db3&platform=android&hash=dd2a0fff6b00ed3ac7ca7070d20360e3' isn't readable.]

--- a/apps/native-component-list/src/screens/FileSystemScreen.tsx
+++ b/apps/native-component-list/src/screens/FileSystemScreen.tsx
@@ -308,6 +308,18 @@ export default class FileSystemScreen extends React.Component<object, State> {
     });
   };
 
+  _downloadAndReadLocalAsset = async () => {
+    const asset = Asset.fromModule(require('../../assets/index.html')).uri;
+    const tmpFile = FileSystem.cacheDirectory + 'test.html';
+    try {
+      await FileSystem.downloadAsync(asset, tmpFile);
+      const result = await FileSystem.readAsStringAsync(tmpFile);
+      Alert.alert('Result', result);
+    } catch (e) {
+      Alert.alert('Error', e.message);
+    }
+  };
+
   render() {
     return (
       <ScrollView style={{ padding: 10 }}>
@@ -334,6 +346,7 @@ export default class FileSystemScreen extends React.Component<object, State> {
         <ListButton onPress={this._getInfoAsset} title="Get Info Asset" />
         <ListButton onPress={this._copyAndReadAsset} title="Copy and Read Asset" />
         <ListButton onPress={this._alertFreeSpace} title="Alert free space" />
+        <ListButton onPress={this._downloadAndReadLocalAsset} title="Download & local asset" />
         {Platform.OS === 'android' && (
           <>
             <HeadingText>Storage Access Framework</HeadingText>

--- a/apps/native-component-list/src/screens/FileSystemScreen.tsx
+++ b/apps/native-component-list/src/screens/FileSystemScreen.tsx
@@ -346,7 +346,10 @@ export default class FileSystemScreen extends React.Component<object, State> {
         <ListButton onPress={this._getInfoAsset} title="Get Info Asset" />
         <ListButton onPress={this._copyAndReadAsset} title="Copy and Read Asset" />
         <ListButton onPress={this._alertFreeSpace} title="Alert free space" />
-        <ListButton onPress={this._downloadAndReadLocalAsset} title="Download and read local asset" />
+        <ListButton
+          onPress={this._downloadAndReadLocalAsset}
+          title="Download and read local asset"
+        />
         {Platform.OS === 'android' && (
           <>
             <HeadingText>Storage Access Framework</HeadingText>

--- a/apps/native-component-list/src/screens/FileSystemScreen.tsx
+++ b/apps/native-component-list/src/screens/FileSystemScreen.tsx
@@ -346,7 +346,7 @@ export default class FileSystemScreen extends React.Component<object, State> {
         <ListButton onPress={this._getInfoAsset} title="Get Info Asset" />
         <ListButton onPress={this._copyAndReadAsset} title="Copy and Read Asset" />
         <ListButton onPress={this._alertFreeSpace} title="Alert free space" />
-        <ListButton onPress={this._downloadAndReadLocalAsset} title="Download & local asset" />
+        <ListButton onPress={this._downloadAndReadLocalAsset} title="Download and read local asset" />
         {Platform.OS === 'android' && (
           <>
             <HeadingText>Storage Access Framework</HeadingText>

--- a/packages/expo-file-system/CHANGELOG.md
+++ b/packages/expo-file-system/CHANGELOG.md
@@ -10,6 +10,7 @@
 
 - On `iOS`, set `httpMethod` on upload requests. ([#26516](https://github.com/expo/expo/pull/26516) by [@alanjhughes](https://github.com/alanjhughes))
 - On `iOS`, fix upload task requests. ([#26880](https://github.com/expo/expo/pull/26880) by [@alanjhughes](https://github.com/alanjhughes))
+- [iOS] Fix downloadAsync for local files. ([#27187](https://github.com/expo/expo/pull/27187) by [@gabrieldonadel](https://github.com/gabrieldonadel))
 
 ### 💡 Others
 

--- a/packages/expo-file-system/ios/FileSystemModule.swift
+++ b/packages/expo-file-system/ios/FileSystemModule.swift
@@ -137,20 +137,20 @@ public final class FileSystemModule: Module {
       if sourceUrl.isFileURL {
         try ensurePathPermission(appContext, path: sourceUrl.path, flag: .read)
         EXFileSystemLocalFileHandler.copy(from: sourceUrl, to: localUrl, resolver: promise.resolver, rejecter: promise.legacyRejecter)
-      } else {
-        let session = options.sessionType == .background ? backgroundSession : foregroundSession
-        let request = createUrlRequest(url: sourceUrl, headers: options.headers)
-        let downloadTask = session.downloadTask(with: request)
-        let taskDelegate = EXSessionDownloadTaskDelegate(
-          resolve: promise.resolver,
-          reject: promise.legacyRejecter,
-          localUrl: localUrl,
-          shouldCalculateMd5: options.md5
-        )
-
-        sessionTaskDispatcher.register(taskDelegate, for: downloadTask)
-        downloadTask.resume()
+        return
       }
+      let session = options.sessionType == .background ? backgroundSession : foregroundSession
+      let request = createUrlRequest(url: sourceUrl, headers: options.headers)
+      let downloadTask = session.downloadTask(with: request)
+      let taskDelegate = EXSessionDownloadTaskDelegate(
+        resolve: promise.resolver,
+        reject: promise.legacyRejecter,
+        localUrl: localUrl,
+        shouldCalculateMd5: options.md5
+      )
+
+      sessionTaskDispatcher.register(taskDelegate, for: downloadTask)
+      downloadTask.resume()
     }
 
     AsyncFunction("uploadAsync") { (targetUrl: URL, localUrl: URL, options: UploadOptions, promise: Promise) in

--- a/packages/expo-file-system/ios/FileSystemModule.swift
+++ b/packages/expo-file-system/ios/FileSystemModule.swift
@@ -134,18 +134,23 @@ public final class FileSystemModule: Module {
       try ensureFileDirectoryExists(localUrl)
       try ensurePathPermission(appContext, path: localUrl.path, flag: .write)
 
-      let session = options.sessionType == .background ? backgroundSession : foregroundSession
-      let request = createUrlRequest(url: sourceUrl, headers: options.headers)
-      let downloadTask = session.downloadTask(with: request)
-      let taskDelegate = EXSessionDownloadTaskDelegate(
-        resolve: promise.resolver,
-        reject: promise.legacyRejecter,
-        localUrl: localUrl,
-        shouldCalculateMd5: options.md5
-      )
+      if sourceUrl.isFileURL {
+        try ensurePathPermission(appContext, path: sourceUrl.path, flag: .read)
+        EXFileSystemLocalFileHandler.copy(from: sourceUrl, to: localUrl, resolver: promise.resolver, rejecter: promise.legacyRejecter)
+      } else {
+        let session = options.sessionType == .background ? backgroundSession : foregroundSession
+        let request = createUrlRequest(url: sourceUrl, headers: options.headers)
+        let downloadTask = session.downloadTask(with: request)
+        let taskDelegate = EXSessionDownloadTaskDelegate(
+          resolve: promise.resolver,
+          reject: promise.legacyRejecter,
+          localUrl: localUrl,
+          shouldCalculateMd5: options.md5
+        )
 
-      sessionTaskDispatcher.register(taskDelegate, for: downloadTask)
-      downloadTask.resume()
+        sessionTaskDispatcher.register(taskDelegate, for: downloadTask)
+        downloadTask.resume()
+      }
     }
 
     AsyncFunction("uploadAsync") { (targetUrl: URL, localUrl: URL, options: UploadOptions, promise: Promise) in

